File: app/src/main/java/gt/hack/nfc/fragment/CheckinFlowFragment.java
package gt.hack.nfc.fragment;


import android.nfc.FormatException;
import android.nfc.NdefMessage;
import android.nfc.NdefRecord;
import android.nfc.NfcAdapter;
import android.nfc.Tag;
import android.nfc.tech.Ndef;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.annotation.Nullable;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.support.v7.widget.AppCompatButton;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import java.io.IOException;
import java.util.HashMap;
import java.util.concurrent.ExecutionException;

import gt.hack.nfc.R;
import gt.hack.nfc.util.CheckInAsyncTask;
import gt.hack.nfc.util.Util;

public class CheckinFlowFragment extends Fragment {
    public static int READER_FLAGS =
            NfcAdapter.FLAG_READER_NFC_A;
    private String id;
    private String name;
    private String email;
    private String school;
    private String branch;
    private String confirmBranch;
    private boolean alreadyCheckedIn = false;
    private boolean wroteBadge = false;
    private final String packageName = "gt.hack.nfc";
    private AppCompatButton confirmButton;

    public static CheckinFlowFragment newInstance(UserFragment user) {
        CheckinFlowFragment f = new CheckinFlowFragment();
        // Supply index input as an argument.
        Bundle args = new Bundle();
        args.putString("id", user.id);
        args.putString("name", user.name);
        args.putString("email", user.email);
        String school = Util.getValueOfQuestion(user.questions, "school");
        args.putString("school", school);

        if (user.application != null) {
            args.putString("branch", user.application.type);
        }

        if (user.confirmation != null) {
            args.putString("confirmBranch", user.confirmation.type);
        }

        f.setArguments(args);

        return f;
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater,
                             @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        Bundle bundle = this.getArguments();
        id = bundle.getString("id");
        name = bundle.getString("name");
        email = bundle.getString("email");
        school = bundle.getString("school");
        branch = bundle.getString("branch");
        confirmBranch = bundle.getString("confirmBranch");
        return inflater.inflate(R.layout.fragment_checkin_confirm, container, false);
    }

    @Override
    public void onResume() {
        super.onResume();
        final ProgressBar progressBar = getActivity().findViewById(R.id.waitForBadge);

        TextView nameView = getActivity().findViewById(R.id.hacker_checkin_name);
        nameView.setText(name);

        TextView emailView = getActivity().findViewById(R.id.hacker_checkin_email);
        emailView.setText(email);

        TextView schoolView = getView().findViewById(R.id.hacker_checkin_school);
        if (school != null) {
            schoolView.setText(school);
        }

        TextView branchView = getView().findViewById(R.id.hacker_checkin_type);
        if (branch != null) {
            branchView.setText(branch);
        }

        TextView confirmBranchView = getView().findViewById(R.id.hacker_confirm_type);
        if (confirmBranch != null) {
            confirmBranchView.setText(confirmBranch);
        }

        confirmButton = getActivity().findViewById(R.id.confirmCheckin);
        confirmButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                try {
                    HashMap<String, TagFragment> tags = new CheckInAsyncTask(
                            PreferenceManager.getDefaultSharedPreferences(getActivity()),
                            CheckinFlowFragment.this).execute(id, "hackgt").get();
                    if (tags != null && tags.get("hackgt").checked_in) {
                        getFragmentManager().popBackStack();
                    } else {
                        Util.makeSnackbar(getActivity().findViewById(R.id.content_frame),
                                R.string.checkin_error, Snackbar.LENGTH_SHORT).show();
                    }
                } catch (InterruptedException e) {
                    e.printStackTrace();
                    Util.makeSnackbar(getActivity().findViewById(R.id.content_frame),
                            R.string.checkin_error, Snackbar.LENGTH_SHORT).show();
                } catch (ExecutionException e) {
                    e.printStackTrace();
                    Util.makeSnackbar(getActivity().findViewById(R.id.content_frame),
                            R.string.checkin_error, Snackbar.LENGTH_SHORT).show();
                }
            }
        });
        if (alreadyCheckedIn) {
            confirmButton.setText("User already checked in");
            confirmButton.setEnabled(false);
            confirmButton.setVisibility(View.VISIBLE);
            progressBar.setVisibility(View.GONE);
            TextView nfcInstructions = getActivity().findViewById(R.id.nfcInstructions);
            nfcInstructions.setVisibility(View.GONE);
        } else {
            NfcAdapter nfc = NfcAdapter.getDefaultAdapter(getActivity());
            if (nfc != null) {
                nfc.enableReaderMode(getActivity(), new NfcAdapter.ReaderCallback() {
                    @Override
                    public void onTagDiscovered(Tag tag) {
                        Ndef ndef = Ndef.get(tag);
                        Log.d("NFC", tag.toString());
                        try {
                            ndef.connect();
                            if (ndef.isWritable() && !wroteBadge) {

                                String type = "badge";

                                NdefRecord uriRecord = NdefRecord.createUri(
                                        "https://live.hack.gt/?user=" + id);
                                NdefMessage ndefMessage = new NdefMessage(
                                        new NdefRecord[] { uriRecord });

                                ndef.writeNdefMessage(ndefMessage);
                                // Uncomment to make the tag read-only in production
                                //ndef.makeReadOnly();

                                getActivity().runOnUiThread(new Runnable() {
                                    public void run() {
                                        progressBar.setVisibility(View.GONE);
                                        ImageView check = getActivity()
                                                .findViewById(R.id.badgeWritten);
                                        check.setVisibility(View.VISIBLE);
                                        confirmButton.setVisibility(View.VISIBLE);
                                    }
                                });
                                wroteBadge = true;
                            }
                            else if (!ndef.isWritable()) {
                                // Tag already locked or unwritable NFC device like a Buzzcard was tapped
                                Util.makeSnackbar(getActivity().findViewById(R.id.content_frame), R.string.unwritable_tag, Snackbar.LENGTH_SHORT).show();
                            }
                        } catch (IOException | FormatException e) {
                            e.printStackTrace();
                        } finally {
                            try {
                                ndef.close();
                            } catch (IOException e) {
                                e.printStackTrace();
                            }
                        }
                    }
                }, READER_FLAGS, null);
            }
        }
    }
}

File: app/src/main/java/gt/hack/nfc/fragment/CheckinFlowFragment.kt
package gt.hack.nfc.fragment

import android.nfc.FormatException
import android.nfc.NdefMessage
import android.nfc.NdefRecord
import android.nfc.NfcAdapter
import android.nfc.tech.Ndef
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.TextView
import androidx.preference.PreferenceManager
import com.google.android.material.snackbar.Snackbar
import gt.hack.nfc.BuildConfig
import gt.hack.nfc.R
import gt.hack.nfc.util.API
import gt.hack.nfc.util.NFCHandler
import gt.hack.nfc.util.Util
import kotlinx.android.synthetic.main.fragment_checkin_confirm.*
import kotlinx.coroutines.runBlocking
import java.io.IOException
import java.util.concurrent.ExecutionException

class CheckinFlowFragment : androidx.fragment.app.Fragment() {
  private var uuid: String? = null
  private var name: String? = null
  private var email: String? = null
  private var school: String? = null
  private var branch: String? = null
  private var confirmBranch: String? = null
  private var alreadyCheckedIn = false
  private var wroteBadge = false
  private val nfcHandler = NFCHandler()
  private val TAG = "checkin2/CHECKIN_FLOW"

  override fun onCreateView(inflater: LayoutInflater,
                            container: ViewGroup?, savedInstanceState: Bundle?): View? {
    super.onCreateView(inflater, container, savedInstanceState)
    val bundle = this.arguments
    uuid = bundle!!.getString("id")
    name = bundle.getString("name")
    email = bundle.getString("email")
    school = bundle.getString("school")
    branch = bundle.getString("branch")
    confirmBranch = bundle.getString("confirmBranch")

    return inflater.inflate(R.layout.fragment_checkin_confirm, container, false)
  }

  override fun onResume() {
    super.onResume()
    if (BuildConfig.DEBUG) {
      unlocked_tag_dev_version_icon.show()
      unlocked_tag_dev_version_text.show()
      lock_tag_notice_icon.hide()
      lock_tag_notice_text.hide()
      no_tag_lock_checkbox.hide()
    } else {
      no_tag_lock_checkbox.setOnCheckedChangeListener { _, checked ->
        if (checked) {
          lock_tag_notice_icon.hide()
          lock_tag_notice_text.hide()
          unlocked_tag_warning_icon.show()
          unlocked_tag_warning_text.show()
        } else {
          lock_tag_notice_icon.show()
          lock_tag_notice_text.show()
          unlocked_tag_warning_icon.hide()
          unlocked_tag_warning_text.hide()
        }
      }
    }

    hacker_checkin_name.text = name
    hacker_checkin_email.text = email

    if (school != null) {
      hacker_checkin_school.text = school
    }

    if (branch != null) {
      hacker_checkin_type.text = branch
    }

    if (confirmBranch != null) {
      hacker_confirm_type.text = confirmBranch
    }

    confirmCheckin.setOnClickListener {
      try {
          val preferences = PreferenceManager.getDefaultSharedPreferences(activity)
          val checkinUuid = uuid
          val tagName = "hackgt"
          if (checkinUuid != null) {
            val checkinResult = runBlocking {
              API.checkInTag(preferences, checkinUuid, tagName,
                  true)
            }

            if (checkinResult != null) {
              var checkinDetails: TagFragment? = null
              val checkinTagInfo = checkinResult.tags().findLast { it.fragments().tagFragment().tag.name.equals(tagName) }
              if (checkinTagInfo != null) {
                checkinDetails = checkinTagInfo.fragments().tagFragment()
              }
              val userInfo = checkinResult.user().fragments().userFragment()

              if (checkinDetails != null) {
                if (userInfo.accepted && userInfo.confirmed) {
                  if (checkinDetails.checkin_success) {
                    fragmentManager!!.popBackStack()
                    Util.makeSnackbar(activity!!.findViewById(R.id.content_frame),
                        R.string.checkin_success, Snackbar.LENGTH_SHORT).show()
                  } else {
                    alreadyCheckedIn = true
                    Util.makeSnackbar(activity!!.findViewById(R.id.content_frame),
                        R.string.user_already_checked_in, Snackbar.LENGTH_SHORT).show()
                  }
                } else {
                  Log.i(TAG, "user (uuid: " + checkinUuid + ") not acccepted and confirmed")
                  Util.makeSnackbar(activity!!.findViewById(R.id.content_frame),
                      R.string.user_not_accepted_and_confirmed, Snackbar.LENGTH_SHORT)
                      .show()
                }

              } else {
                Log.i(TAG, "checkin API returned null")
                Util.makeSnackbar(activity!!.findViewById(R.id.content_frame),
                    R.string.checkin_error, Snackbar.LENGTH_SHORT).show()
              }
            }
          }
        } catch (e: InterruptedException) {
          e.printStackTrace()
        Util.makeSnackbar(activity!!.findViewById(R.id.content_frame),
              R.string.checkin_error, Snackbar.LENGTH_SHORT).show()
      } catch (e: ExecutionException) {
          e.printStackTrace()
        Util.makeSnackbar(activity!!.findViewById(R.id.content_frame),
              R.string.checkin_error, Snackbar.LENGTH_SHORT).show()
      }
    }

    // load nfc loading/errors
    if (view != null) {
      loadNFC(view!!)
    }

  }

  private fun shouldLockTag(keepUnlocked: Boolean, isDebug: Boolean): Boolean {
    if (isDebug || keepUnlocked) {
      return false
    }

    return true

  }

  private fun loadNFC(view: View) {
    val nfcInfo = view.findViewById<TextView>(R.id.nfcInstructions)
    val progressBar = view.findViewById<ProgressBar>(R.id.wait_for_badge_tap)
    val warningIcon = view.findViewById<ImageView>(R.id.nfc_error)
    val nfcEnableButton = view.findViewById<Button>(R.id.enable_nfc_button)

    nfcHandler.activity = activity
    nfcHandler.context = context
    nfcHandler.callback = NfcAdapter.ReaderCallback { tag ->
      val ndef = Ndef.get(tag)
      Log.d("NFC", tag.toString())
      try {
        ndef.connect()
        if (ndef.isWritable && !wroteBadge) {

          val type = "badge"

          val uriRecord = NdefRecord.createUri(
              "https://info.hack.gt/?user=" + uuid)
          val ndefMessage = NdefMessage(
              arrayOf(uriRecord))

          try {
            val lockTag = shouldLockTag(no_tag_lock_checkbox.isChecked, BuildConfig.DEBUG)
            ndef.writeNdefMessage(ndefMessage)
            Log.i(TAG, "Should tag be locked? ${lockTag}")

            if (ndef.canMakeReadOnly()) {
              if (lockTag) {
                ndef.makeReadOnly()
              } else if (BuildConfig.DEBUG) {
                Util.makeSnackbar(activity!!.findViewById(R.id.content_frame), R.string.permanent_badge_locking_option_disabled_debug_build, Snackbar.LENGTH_SHORT).show()
              } else {
                Util.makeSnackbar(activity!!.findViewById(R.id.content_frame), R.string.permanent_badge_locking_option_disabled, Snackbar.LENGTH_SHORT).show()
              }
            } else if (!ndef.canMakeReadOnly() && lockTag) {
              Util.makeSnackbar(activity!!.findViewById(R.id.content_frame), R.string.unlockable_tag, Snackbar.LENGTH_SHORT).show()
            }

            activity!!.runOnUiThread {
              wait_for_badge_tap.hide()
              nfcInstructions.hide()
              badgeWritten.show()
              confirmCheckin.show()
              no_tag_lock_checkbox.hide()
            }
            wroteBadge = true
          } catch (e: IOException) {
            Util.makeSnackbar(activity!!.findViewById(R.id.content_frame), R.string.badge_write_error, Snackbar.LENGTH_LONG).show()
          }


        } else if (!ndef.isWritable) {
          // Tag already locked or unwritable NFC device like a Buzzcard was tapped
          Util.makeSnackbar(activity!!.findViewById(R.id.content_frame), R.string.unwritable_tag, Snackbar.LENGTH_LONG).show()
        }
      } catch (e: IOException) {
        e.printStackTrace()
      } catch (e: FormatException) {
        e.printStackTrace()
      } finally {
        try {
          ndef.close()
        } catch (e: IOException) {
          e.printStackTrace()
        }

      }
    }

    nfcHandler.loadNFC(nfcInfo, progressBar, warningIcon, nfcEnableButton)
  }

  companion object {
    var READER_FLAGS = NfcAdapter.FLAG_READER_NFC_A

    fun newInstance(user: UserFragment): CheckinFlowFragment {
      val f = CheckinFlowFragment()
      // Supply index input as an argument.
      val args = Bundle()
      args.putString("id", user.id)
      args.putString("name", user.name)
      args.putString("email", user.email)
      var school = Util.getValueOfQuestion(user.questions, "school")
      if (school.isNullOrBlank()) {
        school = Util.getValueOfQuestion(user.questions, "university")
      }
      args.putString("school", school)

      if (user.application != null) {
        args.putString("branch", user.application.type)
      }

      if (user.confirmation != null) {
        args.putString("confirmBranch", user.confirmation.type)
      }

      f.arguments = args

      return f
    }
  }

  private fun View.show() {
    this.visibility = View.VISIBLE
  }

  private fun View.hide() {
    this.visibility = View.GONE
  }
}
